File: leet-code/backspaceCompare.js
/**
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function (S, T) {
  return back(S) === back(T)
}

function back (str) {
  let stack = []

  for (let i = 0; i < str.length; i++) {
    const element = str[i]
    if (element === '#') {
      stack.pop()
      continue
    }
    stack.push(element)
  }
  return stack.join('')
}

console.log(backspaceCompare('a#b#c', 'a#d#c'))

File: leet-code/backspaceCompare.go
package main

func backspaceCompare(S string, T string) bool {
	var back func(str string) string
	back = func(str string) string {
		stack := make([]rune, 0)
		for i, v := range str {
			if str[i] == '#' {
				if len(stack) == 0 {
					continue
				}
				stack = stack[:len(stack)-1]
				continue
			}
			stack = append(stack, v)
		}
		return string(stack)
	}

	return back(S) == back(T)
}
